proposals/channel-jamming-mitigation.md

```diff
@@ -0,0 +1,257 @@
+a `channel_update` messages where upfront fees are > 10% of their success-case
+equivalent. This eliminates the risk of nodes setting these fees so high that
+it becomes economically rational for forwarding nodes to fail payments for the
+sake of their upfront fees. 
+
+The upfront fee amount should simply be assigned to the receiving party's 
+balance when an incoming HTLC is added. These amounts are not expected to be
+enforceable on-chain (as they are likely to be dust), so there is no need to 
+include an output for them in the channel's commitment transactions.
+
+As with success-case fees, upfront fees are charged on the outgoing link and 
+accumulated on the incoming link when payments are accepted for forwarding. The
+`update_add_htlc` message is extended with a new TLV which specifies the amount
+that must be pushed to the remote peer on addition of the incoming HTLC to its
+channel state. 
+```
+1. `tlv_stream`: `update_add_htlc_tlvs`
+2. types:
+    1. type: 2 (`upfront_fee_msat`)
+    2. data:
+        * [`tu64`:`upfront_fee_msat`]
+```
+
+The upfront fee that should be sent to the next peer in the route is provided
+in the onion's payload. As is currently the case with htlc forwarding amounts, 
+the difference between the incoming `upfront_fee_msat` amount in 
+`update_add_htlc_tlvs` and the outgoing `upfront_fee_to_forwad` is used to 
+validate that sufficient upfront fees are paid to the forwarding node. 
+```
+1. `tlv_stream`: `payload`
+1. type: 2 (`upfront_fee_to_forward`)
+    2. data:
+        * [`tu64`:`upfront_fee_to_forward`]
+```
+
+An upfront fee policy field is added to bolt-11 invoices to allow receiving 
+nodes to advertise the upfront fees that it will accept for the final hop 
+(since there is no outgoing link for the sender to obtain a policy from). 
+```
+`u` (17) `data_length` 26:
+ * `upfront_base_msat` (32 bits, big-endian)
+ * `upfront_proportional_millionths` (32 bits, big-endian)
+```
+
+Adding a field to the invoice is chosen rather than allowing sender to select
+an arbitrary upfront fee for the final hop, because they have little incentive
+to protect the receiver's privacy at their own expense. The receiver must 
+include the upfront fees paid across all HTLCs in a set as contributions to 
+the total payment amount. This allows receivers to set privacy-preserving 
+policies without shifting the cost to the sender.
+
+
+For example, consider an invoice for 30,000 msat which is paid in multiple 
+parts:
+- HTLC 1 arrives: amount = 15,000, upfront_fees=100
+- HTLC 1 fails with MPPTimeout
+- HTLC 2 arrives: amount = 15,000, upfront_fees=100
+- HTLC 3 arrives: amount = 14,700, upfront_fees=100
+
+The receiving node should reveal the preimage at this point, because it has
+received the total of 30,000 msat from the sender (29,700 paid via HTLCs in the
+set and 300 msat through upfront fees).
+
+Sending nodes can factor this upfront fee into the total amount they dispatch:
```

Comment:
One thing is that with htlcs in-flight, the sender doesn't know yet whether the htlc will reach the final destination or not. If it does, the upfront fee counts towards the total amount. If it doesn't they just pay intermediate nodes do not increase the total amount at the receiver. Maybe this complicates implementation.